### ISSUE TITLE
[Request] Don't install modules at runtime

### ISSUE THREAD

**Issue by kaysond** [Ylianst/MeshCentral#5779]
## Background
I just got some micro desktops with vPro support, and I wanted to install MeshCentral using the Docker image in [uldiseihenbergs/meshcentral-builds](https://github.com/uldiseihenbergs/meshcentral-builds/). A best practice for Docker containers is to run services as a unique user (particularly, non-root), which can be accomplished with `--user` or `user:` in a docker compose file. Unfortunately, doing that didn't work because of a permissions issue with "module installation".

I dug into the code and discovered that dependencies are being gathered and `npm install` is being executed by the main script at runtime:
https://github.com/Ylianst/MeshCentral/blob/990da39fdeb484d243b83c0e4a626a4e6a96a834/meshcentral.js#L4019-L4107

### Permissions
The above means that the permissions issue in Docker can never really be resolved. The image must be created with a particular user selected at build time, and that generally should be root. The container should then ideally be able to run as any user, but in order to be able do an `npm install` at run time, the build user would have to be chosen in advance to be the same as the run user. (Or you could grant everyone write permissions, but that's not a great choice for security).

### Non-ephemeral container
There's also a separate but related issue when it comes to Docker, and that is that you should not write directly to the container's writable layer because 1) the container is ephemeral and should contain everything needed to run the application, and 2) for performance reasons. You can get around this with a volume, but then you run into issues with volume persistence in addition to violating Docker image design principles.

## Dangers With Runtime Module Installation
I did a lot of searching, and I couldn't find any other npm packages, services, etc, that install modules at runtime in this way. From my own experience, and according to the nodejs docs, this is a bad practice. It's fragile in that you have to maintain the main dependency list in two places (`package.json` and `meshcentral.js`), and from looking through the issues it seems that module installation issues come up frequently. The whole point of npm is to resolve dependencies and install packages from a `package.json` (really: `package-lock.json`). It's made to take care of this for you, and if you properly define all of your dependencies I would guess you'll get far fewer reports of issues with installation and update because all of the versioning is frozen and reproducible.

## Disk Space
I'm curious why it was done this way. I'm guessing it's to save the disk space of modules that aren't actually used by the config, and I am seeing 21MB on a "default" install, and 357MB with all modules installed, and 100MB of that is googleapis. Nonetheless, there's a danger in modifying your "executable" based on user configuration, and I don't think even the full 357MB is so concerning that it's worth trying to save.

## Request
I think the project (and Docker users!) would benefit from moving all of the dependencies into `package.json` and removing the runtime install code. **I'd also be happy to contribute the change, including documentation updates, etc.** Another benefit of doing this is it makes installing and running meshcentral trivial with [npx](https://docs.npmjs.com/cli/v7/commands/npx/)

Bonus: I'd be happy to contribute an official docker container too, including github actions to automatically publish it.

### The exception to the rule
npm doesn't support os-specific dependencies, so you can't add windows-only packages to `dependencies` or Unix systems will fail installation. You would add it to `optionalDependencies`, where the failure won't prevent Unix/Mac installs, but windows users should get the packages. There is some risk that those packages fail to install on windows for some other reason, so a runtime check might be warranted.

**Comment by kaysond** [comment]
PS: Thanks for reading my essay :)

**Comment by si458** [comment]
https://github.com/uldiseihenbergs/meshcentral-builds/ is an unofficial docker image...
Nothing to do with meshcentral

Our official one is listed on the main repo under packages right hand sidehttps://github.com/Ylianst/MeshCentral/pkgs/container/meshcentral

All the docker code is here https://github.com/Ylianst/MeshCentral/tree/master/docker

**Comment by si458** [comment]
If you think u can improve our docker image then please have at it!
More than happy to accept PR requests...

**Comment by Jamesits** [comment]
This is still an issue for both container image users. To elaborate:
- The container have to be run as root (non-root containers are enforced by a lot companys) (e.g. https://github.com/Ylianst/MeshCentral/issues/5545 )
- Supply chain attack surface (no way to hash-check the NPM packages)
- The container image does not work offline at all (e.g. https://github.com/Ylianst/MeshCentral/issues/6568 )
- Additional network traffic on each container restart
- Slow start time

Plus additional trouble for the project maintainers:
- You have to remember to update package versions in `meshcentral.js` and `Dockerfile` and `package.json`, and they have to match, otherwise you get undefined/untested behaviors (e.g. https://github.com/Ylianst/MeshCentral/issues/6459 https://github.com/Ylianst/MeshCentral/issues/5640 https://github.com/Ylianst/MeshCentral/issues/4471 https://github.com/Ylianst/MeshCentral/issues/2512 https://github.com/Ylianst/MeshCentral/issues/1957 https://github.com/Ylianst/MeshCentral/issues/5694 ...)

The solution: just add all the packages into `package.json` and generate a lock file with the package manager of your choice! Then remove the package checks completely from `meshcentral.js`.

> you can't add windows-only packages to dependencies or Unix systems will fail installation. 

All the Windows-only packages correctly installs under Linux and macOS! I've tested that. If you want to try for yourself, you can clone `https://github.com/Jamesits/docker-meshcentral2.git` on your OS and do a `pnpm install` (outside Docker).

> for example someone ran the docker image with syslogtcp which required python in the docker image…
I would like to try creating separate images at some point mc/mongodb mc/mysql mc/python etc if that makes sense

No, this is not the way to do it. Your container image should only contain everything your program need to run. For example, if you need to call mongodb client tools or other database client for daily backup (a function of your program), then include them. `syslogtcp` or even `python` is not of one of these cases.

If an user want to collect syslog of a container, they are better off doing it in a way that is non-intrusive. There are a lot of ways to achieve this. One common way is to [configure an alternative logging driver](https://docs.docker.com/engine/logging/drivers/syslog/) on their container runtime; another common choise is to [use a sidecar container](https://medium.com/@bindubc/the-sidecar-pattern-2317721916bf).

If an user really need a certain tool they use inside the container, they can simply add layers to your container with the following Dockerfile:
```dockerfile
FROM your-image:tag
RUN ...
```
This is cleaner and more managable than tell the upstream packager to include the universe in their container image. Plus, large container images is not a problem. OCI container images are designed to share layers with each other to reduce disk usage, and ~1GiB is not an extreme size for such a container.

---

Besides that, [MeshCentral tries to find its config files relative to its code](https://github.com/Ylianst/MeshCentral/blob/b34d314cd29d79136177d09326c57d4d266203e3/meshcentral.js#L85) which does not work if the `node_modules` is not a flatten one. Not all npm package manager create a flattened `node_modules`, and an NPM package should not change anything inside the `node_modules`. (e.g. https://github.com/Ylianst/MeshCentral/issues/4318 )

---

The way that the official container image being constructed has brought much trouble to me, so much that I had to maintain [my own container image for MeshCentral](https://github.com/Jamesits/docker-meshcentral2). Here's my approach, for reference:
- Toolchain: corepack + pnpm (just what I'm most familiar with)
- All packages locked and hash-checked during a CI container build
- A script to extract the dependency packages automatically (albeit not the best way to do it...)
- Upstream updates trigger auto PRs by Renovate bot

> More than happy to accept PR requests...

I'd like to contribute, but the changes required exceed what I define as a small progressive patch (small enough so that it has a high chance of being reviewed even if submitted by a random person). And it's the engineering choices I mentioned above that is proactively preventing people from contributing by making any change unneedingly expensive. The same engineering choices lead to a lot trouble and confusion to other users over the years, no matter if they use MeshCentral inside a container or not.

(Don't know how to write an ending for this comment. Sorry.)

**Comment by DaanSelen** [comment]
Is this still an issue?

**Comment by Jamesits** [comment]
@DaanSelen Yes, it is still an issue.

**Comment by DaanSelen** [comment]
And assuming this is also the case for the PR in progress? https://github.com/Ylianst/MeshCentral/pull/6937

**Comment by Jamesits** [comment]
@DaanSelen Thank you! It seems that PR does not solve the original problem described by this issue (dynamic node package installation at runtime) though?

**Comment by si458** [comment]
im not sure how people would like us to resolve this issue?
we cant include every single npm package in the package.json
because the size/package count of meshcentral would get silly and out of control!
as is at the moment the new docker image size is 1GB and that includes the 350MB JUST FOR MESHCENTRAL CODE!!!
@DaanSelen did an amazing job of changing the docker image so it now includes database npms by default for everyone
but again if u want certin feature it still needs to install the missing packages
at least if you do a normal npm install its only once it needs to do this!
but with docker because the image is expendable, it has to run it every time im afraid

**Comment by Jamesits** [comment]
What's wrong with 350MB of nodejs code? I mean, if you chose the nodejs ecosystem, you should fully accept its drawbacks, like the dependency hell and a large runtime. These packages are going to be installed on the user's computers anyway. (The packages are around 250MB actually, if it makes you feel better.)

And most Docker users would like a slight larger size image more than having to manage language-specific dependencies inside the container. Docker is invented exactly for this case. No manual dependency management, just immutable images shipped and ready to run. 

> at least if you do a normal npm install its only once it needs to do this!

No, as Docker images are somewhat immutable, these packages are going to be re-downloaded every time the container is restarted. This is a magnitude worse than including them in the image. (Trying to persist these files comes with another set of different challenges.)

If multiple different database variants are viable, try Docker multistage build with different dependencies packaged to different targets, and distributed with different tag suffixes. This is the equivalent of "downloading the packages you need only". It reduces the image size one user needed to download, but is a little more complex to implement.

**Comment by si458** [comment]
@Jamesits i decided that hopefully before the next release this week, 
im going to add in another docker image, something like `latest-full`
and this will include EVERY SINGLE NPM PACKAGE we use on the project
then if people want to download the full package they can do 👍 
hope this is ok?

its also already on my to-do to add in the multi database tags too
so it would be something like `latest-mysql, latest-postgres, latest-mongodb`
and each will include the npm package that needs for that database!

**Comment by Jamesits** [comment]
Thank you, this would be extra helpful!

**Comment by DaanSelen** [comment]
Having different images with different backends would only increase choice and lean operation!

**Comment by si458** [comment]
so i did an experiment,

base size of `node_modules` using `package.json` currently is only 31MB (18 packages in total)

i then editied the dependencies and added ALL the npm packages we use inside of meshcentral.js (70 packages in total)

i then HAD to install `build-essential` and `make` OS packages because of the syslog module requires them to build

i removed node_modules and ran `npm install` again

the total size of `node_modules` is now, 478MB, thats 15 times bigger!!!

so add that on top of the already 1GB, so the image is now 1.5GB in size!

is this acceptable? as personally thats a ridiculous size!

**Comment by kaysond** [comment]
> i then HAD to install `build-essential` and `make` OS packages because of the syslog module requires them to build

For something like this, you can (and should) run a separate build step in its own container, then copy the results into the final image, so you don't have the packages in it.

> so add that on top of the already 1GB, so the image is now 1.5GB in size!
> 
> is this acceptable? as personally thats a ridiculous size!

Yes it's totally fine. There are plenty of docker images that are even larger. In any case, the users will have to download all the packages anyways. It's best practice to just include them to the image. If you want to reduce the image size, you should reduce the number of dependencies.

**Comment by si458** [comment]
unfortunately we cant reduce the number of dependencies as each feature in meshcentral basically uses a dependency
pushover, discord, telegram, etc for notifications, each have their own package
mysql, mariadb, postgres, mongodo, etc for databases, each have their own package
so we cant just get rid of packages either sadly :(
however, brain wave!
ill speak to @DaanSelen when i get chance as hes maintaining the docker images at the moment for me
we currently have to install python, make, etc in the container for the syslog
we could just do as you explained, install them to build the syslog binarys/nan packages etc,
then just uninstall the python, make, etc as they shouldnt be needed anymore!
i remember @DaanSelen saying something like 300MB was just the python and make/gcc packages!
so they can go haha! thanks for the tip!
sadly i dont think its going to make it in this months release!
but we will get there for next month tho!
we will find a way! ✊

**Comment by kaysond** [comment]
> unfortunately we cant reduce the number of dependencies as each feature in meshcentral basically uses a dependency pushover, discord, telegram, etc for notifications, each have their own package mysql, mariadb, postgres, mongodo, etc for databases, each have their own package so we cant just get rid of packages either sadly :( however, brain wave! ill speak to [@DaanSelen](https://github.com/DaanSelen) when i get chance as hes maintaining the docker images at the moment for me we currently have to install python, make, etc in the container for the syslog we could just do as you explained, install them to build the syslog binarys/nan packages etc, then just uninstall the python, make, etc as they shouldnt be needed anymore! i remember [@DaanSelen](https://github.com/DaanSelen) saying something like 300MB was just the python and make/gcc packages! so they can go haha! thanks for the tip! sadly i dont think its going to make it in this months release! but we will get there for next month tho! we will find a way! ✊

You mentioned python before, not sure if that's what is used in the backend, but have you looked at https://github.com/caronc/apprise ? I imagine there is a similar thing in JS/TS. Similarly, you should look into using an ORM for the databases, if you're not already. Of course, there's always the chance that they are using those dependencies anyways...

**Comment by si458** [comment]
@kaysond we dont use python in ANY which way,  the application is written in NodeJS (Javascript)

some nodejs packages needs to be built to interact with the OS in question for certain APIs
and sadly the syslog module is one of them
so it calls python which calls nan to BUILD the binary for nodejs to run/understand
its annoying as it means each OS u install the package on needs an extra set of build tools
so even windows would need python installed to BUILD the binary

but the only good thing is from my understanding of the nan binarys for nodejs, 
once they are built for that specifc OS, its built, u dont need the python/build-essential again!

so ill do some testing along with @DaanSelen and see what we can come up with
because also in theory the docker images will ALWAYS be linux anyways!

we dont build windows images for docker haha

edit: nodejs example of an add-on which uses nan https://github.com/nodejs/node-addon-examples

edit2: sorry its `modern-syslog` module which uses nan (native binarys) and needs the python to be built
https://www.npmjs.com/package/modern-syslog

**Comment by DaanSelen** [comment]
> unfortunately we cant reduce the number of dependencies as each feature in meshcentral basically uses a dependency pushover, discord, telegram, etc for notifications, each have their own package mysql, mariadb, postgres, mongodo, etc for databases, each have their own package so we cant just get rid of packages either sadly :( however, brain wave! ill speak to [@DaanSelen](https://github.com/DaanSelen) when i get chance as hes maintaining the docker images at the moment for me we currently have to install python, make, etc in the container for the syslog we could just do as you explained, install them to build the syslog binarys/nan packages etc, then just uninstall the python, make, etc as they shouldnt be needed anymore! i remember [@DaanSelen](https://github.com/DaanSelen) saying something like 300MB was just the python and make/gcc packages! so they can go haha! thanks for the tip! sadly i dont think its going to make it in this months release! but we will get there for next month tho! we will find a way! ✊

Yeah the Python and software development libraries take like 300MB, however I am not a JS developer so I am going to look with Simon on importing/removing the relevant modules from the Dockerfile.

**Comment by kaysond** [comment]
Not stale


-------- Original Message --------

**Comment by DaanSelen** [comment]
Like this: https://github.com/Ylianst/MeshCentral/pull/7222 ?

**Comment by kaysond** [comment]
> Like this: [#7222](https://github.com/Ylianst/MeshCentral/pull/7222) ?

Maybe I'm missing something, but this just splits different dependency sets into different image tags. I'm not seeing anything that would install the node packages at build time

**Comment by DaanSelen** [comment]
> > Like this: [#7222](https://github.com/Ylianst/MeshCentral/pull/7222) ?
> 
> Maybe I'm missing something, but this just splits different dependency sets into different image tags. I'm not seeing anything that would install the node packages at build time

Then maybe this one? https://github.com/Ylianst/MeshCentral/pull/7227

**Comment by kaysond** [comment]
I mean it looks like you're pre-installing some libraries:

https://github.com/DaanSelen/MeshCentral/blob/512157a77c29313e6b426008f86b78d4a9890f0e/docker/Dockerfile#L132-L142

but the version is hardcoded in the Dockerfile instead of using package-lock.json which is a bad idea.

And it seems there are still functions in the main `meshcentral.js` to install modules...
https://github.com/DaanSelen/MeshCentral/blob/512157a77c29313e6b426008f86b78d4a9890f0e/meshcentral.js#L4112-L4188

**Comment by kaysond** [comment]
I wish I had more time to help with this, but ultimately, the only time any packages should be installed is in the Dockerfile, and it should just be an `npm install` (after copying in the app files so you have package-lock.json)

**Comment by DaanSelen** [comment]
> I mean it looks like you're pre-installing some libraries:
> 
> https://github.com/DaanSelen/MeshCentral/blob/512157a77c29313e6b426008f86b78d4a9890f0e/docker/Dockerfile#L132-L142
> 
> but the version is hardcoded in the Dockerfile instead of using package-lock.json which is a bad idea.
> 
> And it seems there are still functions in the main `meshcentral.js` to install modules... https://github.com/DaanSelen/MeshCentral/blob/512157a77c29313e6b426008f86b78d4a9890f0e/meshcentral.js#L4112-L4188

True, I would need @si458 's opinion on how to handle JS things that way.
